ldp/alg/callbacks.py

```diff
@@ -123,54 +122,64 @@ async def after_update(self) -> None:
 
 
 class StoreTrajectoriesCallback(Callback):
-    """Simple callback that stores train/eval trajectories in an in-memory list."""
-
-    def __init__(self):
-        self.train_trajectories = []
-        self.eval_trajectories = []
-
-    async def after_train_step(self, trajectories: Sequence[Trajectory]) -> None:
-        self.train_trajectories.extend(trajectories)
-
-    async def after_eval_step(self, trajectories: Sequence[Trajectory]) -> None:
-        self.eval_trajectories.extend(trajectories)
-
-
-class TrajectoryFileCallback(Callback):
     """Callback that writes trajectories to a file."""
 
-    def __init__(self, output_dir: os.PathLike | str):
-        self.output_dir = Path(output_dir)
-        self.output_dir.mkdir(parents=True, exist_ok=True)
+    @staticmethod
+    def default_serialize_env(env: Environment, transition: Transition) -> str | None:
+        if not transition.done:
+            return None
+        try:
+            frame = env.export_frame()
+        except NotImplementedError:
+            return None  # Allow for envs that didn't implement export_frame()
+        return frame.model_dump_json(exclude={"state"}, indent=2)
 
-        self.out_files: dict[str, Path] = {}
+    def __init__(
+        self,
+        output_dir: str | os.PathLike | None = None,
```

Comment:
I think we can force the user to pass `output_dir`. It looks like we need it to be optional to support the `StoreTrajectoriesCallback` functionality, but I don't think anyone is actually using it (just grepped the codebase) - let's deprecate it and simplify this callback

Reply:
So we use `StoreTrajectoriesCallback` in unit tests to in-memory store trajectories. I don't want unit tests to have to use `tmpfile`s when really the tests just want to temporarily house `Trajectory` in-memory.

I would rather have it be optional, because people may not always have a filesystem

Reply:
I see your point re: tests, but:
> I would rather have it be optional, because people may not always have a filesystem
I would contend there isn't a use-case for this callback if you don't have a filesystem - all it'd do is aggregate things into a list. You'd have to subclass and add some further post-processing for any real utility. That would be fairly simple to write from scratch and I'm not sure complicating this class is worth saving the user a little boilerplate.

Reply:
If you don't want to add that in this PR, I can open a follow-up with the simplification / modified tests. IMO it's worth keeping core parts of the code simple and easy-to-read.

Reply:
The use case with no filesystem is basically storing parts of the trajectory for post-rollout processing (e.g. evaluating consensus sampling still within the same Python execution).

Since the file I/O can become sizable (especially in longer trajectories), requiring it forces one to incur a perf hit even though one just needs an in-memory cache.

Lastly, I like being able to not worry about or requiring filesystem for DIY scripts and quick unit tests.

---

Let's just leave this debate for the future

Reply:
Idk, it still feels like we're trying to put two (three?) unrelated things in one callback: aggregating trajectories in a list for another consumer, serializing trajectories to JSONLs, serializing final environment frames to JSONs. 

Happy to defer to another PR, but I do want to keep this part of the codebase simple and easy to read.

Reply:
Sure no problem, we can just revisit this merge later